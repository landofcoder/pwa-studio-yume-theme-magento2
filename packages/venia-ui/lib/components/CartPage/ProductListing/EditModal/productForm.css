.optionRoot {
    border-bottom: 1px solid rgb(var(--venia-global-color-border));
    padding: 1rem 0;
}

.quantityLabel {
    font-size: 1rem;
    font-weight: 600;
    line-height: 1.5;
    margin-bottom: 1.5rem;
    padding-top: 1rem;
}

.quantityRoot {
    composes: root from '../quantity.css';
    grid-template-columns: auto 1fr auto;
    padding: 0 1rem;
}

.submit {
    display: flex;
    justify-content: center;
    padding: 2rem;
}

.loading {
    composes: root from '../../../LoadingIndicator/indicator.css';
    height: unset;
    text-align: center;
}

.dataError {
    color: rgb(var(--venia-global-color-error));
    display: inline-block;
    padding-top: 1rem;
    line-height: var(--venia-global-lineHeight-300);
}

.errorContainer {
    padding-top: 1rem;
}
