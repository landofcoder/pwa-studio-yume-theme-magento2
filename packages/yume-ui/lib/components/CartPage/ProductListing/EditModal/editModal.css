.root {
    background-color: white;
    bottom: 0;
    display: grid;
    grid-template-rows: auto 1fr;
    height: 100%;
    max-width: 360px;
    opacity: 0;
    overflow: hidden;
    position: fixed;
    right: 0;
    top: 0;
    transform: translate3d(100%, 0, 0);
    transition-duration: 192ms;
    transition-timing-function: var(--venia-global-anim-out);
    transition-property: opacity, transform, visibility;
    visibility: hidden;
    width: 100%;
    z-index: 3;
}

.root_open {
    composes: root;
    box-shadow: 1px 0 rgb(var(--venia-global-color-border));
    opacity: 1;
    transform: translate3d(0, 0, 0);
    transition-duration: 224ms;
    transition-timing-function: var(--venia-global-anim-in);
    visibility: visible;
}

.body {
    border-bottom: 1px solid rgb(var(--venia-global-color-border));
    overflow: auto;
    padding: 0.5rem 2rem;
}

.header {
    border-bottom: 1px solid rgb(var(--venia-global-color-border));
    display: flex;
    justify-content: space-between;
    padding: 0.875rem;
}

.headerText {
    align-self: center;
    font-weight: 600;
}
