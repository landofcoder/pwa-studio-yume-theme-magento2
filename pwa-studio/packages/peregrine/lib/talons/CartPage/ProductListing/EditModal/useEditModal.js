import { useState } from 'react';

import { useAppContext } from '../../../../context/app';

/**
 * This talon contains logic for a product edit modal used on a cart page.
 * It returns prop data for rendering an interactive modal component.
 *
 * @function
 *
 * @return {EditModalTalonProps}
 *
 * @example <caption>Importing into your project</caption>
 * import { useEditModal } from '@magento/peregrine/lib/talons/CartPage/ProductListing/EditModal/useEditModal';
 */
export const useEditModal = () => {
    const [{ drawer }, { closeDrawer }] = useAppContext();
    const isOpen = drawer === 'product.edit';

    const [variantPrice, setVariantPrice] = useState(null);

    return {
        handleClose: closeDrawer,
        isOpen,
        setVariantPrice,
        variantPrice
    };
};

/** JSDocs type definitions */

/**
 * Object type returned by the {@link useEditModal} talon.
 * It provides props data for rendering an edit modal component.
 *
 * @typedef {Object} EditModalTalonProps
 *
 * @property {function} handleClose Callback function for handling the closing event of the modal.
 * @property {boolean} isOpen True if the modal is open. False otherwise.
 * @property {function} setVariantPrice Function for setting a product's variant price.
 * @property {Object} variantPrice The variant price for a product. See [Money object]{@link https://devdocs.magento.com/guides/v2.4/graphql/product/product-interface.html#Money}.
 */
