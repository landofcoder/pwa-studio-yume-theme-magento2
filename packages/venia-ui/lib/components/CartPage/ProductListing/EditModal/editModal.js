import React from 'react';
import { FormattedMessage } from 'react-intl';
import { X as CloseIcon } from 'react-feather';
import { useEditModal } from '@magento/peregrine/lib/talons/CartPage/ProductListing/EditModal/useEditModal';

import { mergeClasses } from '../../../../classify';
import Icon from '../../../Icon';
import { Portal } from '../../../Portal';
import defaultClasses from './editModal.css';
import ProductDetail from './productDetail';
import ProductForm from './productForm';

/**
 * A child component of the ProductListing component.
 * This component renders an edit modal for a product.
 *
 * @param {Object} props
 * @param {Object} props.item Product to edit.
 * See [productListingFragments.js]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/ProductListing/productListingFragments.js}
 * for a list of properties for this object.
 * @param {Function} props.setIsCartUpdating Function for setting the updating state of the cart.
 * @param {Object} props.classes CSS className overrides.
 * See [editModal.css]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/ProductListing/EditModal/editModal.css}
 * for a list of classes you can override.
 *
 * @returns {React.Element}
 *
 * @example <caption>Importing into your project</caption>
 * import EditModal from "@magento/venia-ui/lib/components/CartPage/ProductListing/EditModal";
 */
const EditModal = props => {
    const { item, setIsCartUpdating } = props;
    const talonProps = useEditModal();
    const { handleClose, isOpen, setVariantPrice, variantPrice } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);
    const rootClass = isOpen ? classes.root_open : classes.root;

    const bodyComponent = item ? (
        <div className={classes.body} key={item.id}>
            <ProductDetail item={item} variantPrice={variantPrice} />
            <ProductForm
                item={item}
                setIsCartUpdating={setIsCartUpdating}
                setVariantPrice={setVariantPrice}
            />
        </div>
    ) : null;

    return (
        <Portal>
            <aside className={rootClass}>
                <div className={classes.header}>
                    <span className={classes.headerText}>
                        <FormattedMessage
                            id={'editModal.headerText'}
                            defaultMessage={'Edit Item'}
                        />
                    </span>
                    <button
                        className={classes.closeButton}
                        onClick={handleClose}
                    >
                        <Icon src={CloseIcon} />
                    </button>
                </div>
                {bodyComponent}
            </aside>
        </Portal>
    );
};

export default EditModal;
