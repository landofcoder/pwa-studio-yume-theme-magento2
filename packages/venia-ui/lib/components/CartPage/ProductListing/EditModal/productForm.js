import React, { Fragment } from 'react';
import { FormattedMessage, useIntl } from 'react-intl';
import { gql } from '@apollo/client';
import { Form } from 'informed';
import { useProductForm } from '@magento/peregrine/lib/talons/CartPage/ProductListing/EditModal/useProductForm';

import { mergeClasses } from '../../../../classify';
import Button from '../../../Button';
import FormError from '../../../FormError';
import LoadingIndicator from '../../../LoadingIndicator';
import Options from '../../../ProductOptions';
import { QuantityFields } from '../quantity';
import defaultClasses from './productForm.css';
import { CartPageFragment } from '../../cartPageFragments.gql';

const ProductForm = props => {
    const { item: cartItem, setIsCartUpdating, setVariantPrice } = props;
    const { formatMessage } = useIntl();
    const talonProps = useProductForm({
        cartItem,
        getConfigurableOptionsQuery: GET_CONFIGURABLE_OPTIONS,
        setIsCartUpdating,
        setVariantPrice,
        updateConfigurableOptionsMutation: UPDATE_CONFIGURABLE_OPTIONS_MUTATION,
        updateQuantityMutation: UPDATE_QUANTITY_MUTATION
    });
    const {
        configItem,
        errors,
        handleOptionSelection,
        handleSubmit,
        isLoading,
        isSaving
    } = talonProps;

    const classes = mergeClasses(defaultClasses, props.classes);

    if (isLoading || isSaving) {
        const message = isLoading
            ? formatMessage({
                  id: 'productForm.fetchingProductOptions',
                  defaultMessage: 'Fetching Product Options...'
              })
            : formatMessage({
                  id: 'productForm.updatingCart',
                  defaultMessage: 'Updating Cart...'
              });
        return (
            <LoadingIndicator
                classes={{
                    root: classes.loading
                }}
            >
                {message}
            </LoadingIndicator>
        );
    }

    if (!configItem) {
        return (
            <span className={classes.dataError}>
                <FormattedMessage
                    id={'productForm.dataError'}
                    defaultMessage={
                        'Something went wrong. Please refresh and try again.'
                    }
                />
            </span>
        );
    }

    return (
        <Fragment>
            <FormError
                classes={{
                    root: classes.errorContainer
                }}
                errors={Array.from(errors.values())}
                scrollOnError={false}
            />
            <Form onSubmit={handleSubmit}>
                <Options
                    classes={{
                        root: classes.optionRoot
                    }}
                    onSelectionChange={handleOptionSelection}
                    options={configItem.configurable_options}
                    selectedValues={cartItem.configurable_options}
                />
                <h3 className={classes.quantityLabel}>
                    <FormattedMessage
                        id={'productForm.quantity'}
                        defaultMessage={'Quantity'}
                    />
                </h3>
                <QuantityFields
                    classes={{
                        root: classes.quantityRoot
                    }}
                    initialValue={cartItem.quantity}
                    itemId={cartItem.id}
                />
                <div className={classes.submit}>
                    <Button priority="high" type="submit">
                        <FormattedMessage
                            id={'productForm.submit'}
                            defaultMessage={'Update'}
                        />
                    </Button>
                </div>
            </Form>
        </Fragment>
    );
};

export default ProductForm;

export const GET_CONFIGURABLE_OPTIONS = gql`
    query productDetailBySku($sku: String) {
        products(filter: { sku: { eq: $sku } }) {
            items {
                id
                sku
                ... on ConfigurableProduct {
                    configurable_options {
                        attribute_code
                        attribute_id
                        id
                        label
                        values {
                            default_label
                            label
                            store_label
                            use_default_value
                            value_index
                            swatch_data {
                                ... on ImageSwatchData {
                                    thumbnail
                                }
                                value
                            }
                        }
                    }
                    variants {
                        attributes {
                            code
                            value_index
                        }
                        product {
                            id
                            price {
                                regularPrice {
                                    amount {
                                        currency
                                        value
                                    }
                                }
                            }
                            sku
                        }
                    }
                }
            }
        }
    }
`;

export const UPDATE_QUANTITY_MUTATION = gql`
    mutation UpdateCartItemQuantity(
        $cartId: String!
        $cartItemId: Int!
        $quantity: Float!
    ) {
        updateCartItems(
            input: {
                cart_id: $cartId
                cart_items: [{ cart_item_id: $cartItemId, quantity: $quantity }]
            }
        ) @connection(key: "updateCartItems") {
            cart {
                id
                ...CartPageFragment
            }
        }
    }
    ${CartPageFragment}
`;

export const UPDATE_CONFIGURABLE_OPTIONS_MUTATION = gql`
    mutation UpdateConfigurableOptions(
        $cartId: String!
        $cartItemId: Int!
        $parentSku: String!
        $variantSku: String!
        $quantity: Float!
    ) {
        addConfigurableProductsToCart(
            input: {
                cart_id: $cartId
                cart_items: [
                    {
                        data: { quantity: $quantity, sku: $variantSku }
                        parent_sku: $parentSku
                    }
                ]
            }
        ) @connection(key: "addConfigurableProductsToCart") {
            cart {
                id
            }
        }

        removeItemFromCart(
            input: { cart_id: $cartId, cart_item_id: $cartItemId }
        ) @connection(key: "removeItemFromCart") {
            cart {
                id
                ...CartPageFragment
            }
        }
    }
    ${CartPageFragment}
`;
