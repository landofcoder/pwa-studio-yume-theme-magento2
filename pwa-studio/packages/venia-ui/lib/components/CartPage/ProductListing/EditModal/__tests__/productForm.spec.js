import React from 'react';
import { createTestInstance } from '@magento/peregrine';
import { useProductForm } from '@magento/peregrine/lib/talons/CartPage/ProductListing/EditModal/useProductForm';

import ProductForm from '../productForm';

jest.mock(
    '@magento/peregrine/lib/talons/CartPage/ProductListing/EditModal/useProductForm'
);
jest.mock('../../../../../classify');
jest.mock('../../../../LoadingIndicator', () => 'LoadingIndicator');
jest.mock('../../../../ProductOptions', () => 'Options');

const mockItem = {
    id: '123',
    quantity: 5,
    configurable_options: ['option1', 'option2']
};

const mockTalonProps = {
    configItem: {
        configurable_options: ['option3', 'option4']
    },
    errors: new Map(),
    handleOptionSelection: jest.fn(),
    handleSubmit: jest.fn(),
    isLoading: false,
    setFormApi: jest.fn()
};

test('renders loading indicator while options are being fetched', () => {
    useProductForm.mockReturnValueOnce({
        isLoading: true
    });

    const tree = createTestInstance(
        <ProductForm item={{}} setIsUpdating={jest.fn()} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders form with data', () => {
    useProductForm.mockReturnValueOnce(mockTalonProps);

    const tree = createTestInstance(
        <ProductForm item={mockItem} setIsUpdating={jest.fn()} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
});

test('renders form errors', () => {
    useProductForm.mockReturnValueOnce({
        ...mockTalonProps,
        errors: new Map([['error', new Error('Form Error')]])
    });

    const tree = createTestInstance(
        <ProductForm item={mockItem} setIsUpdating={jest.fn()} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
});
