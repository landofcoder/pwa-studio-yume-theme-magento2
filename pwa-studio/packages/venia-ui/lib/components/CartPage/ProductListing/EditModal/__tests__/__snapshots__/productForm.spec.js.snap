// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form errors 1`] = `
Array [
  <div
    className="root"
  >
    <span
      className="errorMessage"
    >
      Form Error
    </span>
  </div>,
  <form
    onKeyDown={[Function]}
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <Options
      classes={
        Object {
          "root": "optionRoot",
        }
      }
      onSelectionChange={[MockFunction]}
      options={
        Array [
          "option3",
          "option4",
        ]
      }
      selectedValues={
        Array [
          "option1",
          "option2",
        ]
      }
    />
    <h3
      className="quantityLabel"
    >
      Quantity
    </h3>
    <div
      className="root"
    >
      <label
        className="label"
        htmlFor="123"
      />
      <button
        aria-label="Decrease Quantity"
        className="button_decrement"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="root"
        >
          <svg
            className="icon"
            fill="none"
            height={22}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={22}
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="5"
              x2="19"
              y1="12"
              y2="12"
            />
          </svg>
        </span>
      </button>
      <span
        className="root"
        style={
          Object {
            "--iconsAfter": 0,
            "--iconsBefore": 0,
          }
        }
      >
        <span
          className="input"
        >
          <input
            aria-label="Item Quantity"
            className="input"
            id="123"
            inputMode="numeric"
            min={0}
            name="quantity"
            onBlur={[Function]}
            onChange={[Function]}
            pattern="[0-9]*"
            value={5}
          />
        </span>
        <span
          className="before"
        />
        <span
          className="after"
        />
      </span>
      <p
        className="root"
      />
      <button
        aria-label="Increase Quantity"
        className="button_increment"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span
          className="root"
        >
          <svg
            className="icon"
            fill="none"
            height={20}
            stroke="currentColor"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="2"
            viewBox="0 0 24 24"
            width={20}
            xmlns="http://www.w3.org/2000/svg"
          >
            <line
              x1="12"
              x2="12"
              y1="5"
              y2="19"
            />
            <line
              x1="5"
              x2="19"
              y1="12"
              y2="12"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      className="submit"
    >
      <button
        className="root_highPriority"
        disabled={false}
        type="submit"
      >
        <span
          className="content"
        >
          Update
        </span>
      </button>
    </div>
  </form>,
]
`;

exports[`renders form with data 1`] = `
<form
  onKeyDown={[Function]}
  onReset={[Function]}
  onSubmit={[Function]}
>
  <Options
    classes={
      Object {
        "root": "optionRoot",
      }
    }
    onSelectionChange={[MockFunction]}
    options={
      Array [
        "option3",
        "option4",
      ]
    }
    selectedValues={
      Array [
        "option1",
        "option2",
      ]
    }
  />
  <h3
    className="quantityLabel"
  >
    Quantity
  </h3>
  <div
    className="root"
  >
    <label
      className="label"
      htmlFor="123"
    />
    <button
      aria-label="Decrease Quantity"
      className="button_decrement"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span
        className="root"
      >
        <svg
          className="icon"
          fill="none"
          height={22}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={22}
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
        </svg>
      </span>
    </button>
    <span
      className="root"
      style={
        Object {
          "--iconsAfter": 0,
          "--iconsBefore": 0,
        }
      }
    >
      <span
        className="input"
      >
        <input
          aria-label="Item Quantity"
          className="input"
          id="123"
          inputMode="numeric"
          min={0}
          name="quantity"
          onBlur={[Function]}
          onChange={[Function]}
          pattern="[0-9]*"
          value={5}
        />
      </span>
      <span
        className="before"
      />
      <span
        className="after"
      />
    </span>
    <p
      className="root"
    />
    <button
      aria-label="Increase Quantity"
      className="button_increment"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      <span
        className="root"
      >
        <svg
          className="icon"
          fill="none"
          height={20}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth="2"
          viewBox="0 0 24 24"
          width={20}
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="12"
            x2="12"
            y1="5"
            y2="19"
          />
          <line
            x1="5"
            x2="19"
            y1="12"
            y2="12"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    className="submit"
  >
    <button
      className="root_highPriority"
      disabled={false}
      type="submit"
    >
      <span
        className="content"
      >
        Update
      </span>
    </button>
  </div>
</form>
`;

exports[`renders loading indicator while options are being fetched 1`] = `
<LoadingIndicator
  classes={
    Object {
      "root": "loading",
    }
  }
>
  Fetching Product Options...
</LoadingIndicator>
`;
